vm/src/obj/objint.rs

```diff
@@ -810,79 +786,147 @@ pub fn to_int(vm: &VirtualMachine, obj: &PyObjectRef, base: &BigInt) -> PyResult
+            bytes_to_int(s.as_bytes(), base)
+        }
+        bytes @ PyBytes => {
+            let bytes = bytes.get_value();
+            bytes_to_int(bytes, base)
+        }
+        bytearray @ PyByteArray => {
+            let inner = bytearray.borrow_value();
+            bytes_to_int(&inner.elements, base)
+        }
+        _ => {
+            return Err(
+                vm.new_type_error("int() can't convert non-string with explicit base".to_owned())
+            );
+        }
+    });
+    match opt {
+        Some(int) => Ok(int),
+        None => Err(vm.new_value_error(format!(
+            "invalid literal for int() with base {}: {}",
+            base,
+            vm.to_repr(obj)?,
+        ))),
+    }
+}
+
+fn bytes_to_int(lit: &[u8], mut base: u32) -> Option<BigInt> {
+    // split sign
+    let mut lit = lit.trim();
+    let sign = match lit.first()? {
+        b'+' => Some(Sign::Plus),
+        b'-' => Some(Sign::Minus),
+        _ => None,
+    };
+    if sign.is_some() {
+        lit = &lit[1..];
+    }
+
+    // split radix
+    let first = *lit.first()?;
+    let has_radix = if first == b'0' {
+        match base {
+            0 => {
+                if let Some(parsed) = lit.get(1).and_then(detect_base) {
+                    base = parsed;
+                    true
+                } else {
+                    if let Some((last, body)) = lit[1..].split_last() {
+                        let is_zero = body.iter().all(|&c| c == b'0' || c == b'_') && *last == b'0';
+                        if !is_zero {
+                            return None;
+                        }
+                    }
+                    return Some(BigInt::zero());
                 }
-            } else if (matched_radix == 2 && base < 12)
-                || (matched_radix == 8 && base < 25)
-                || (matched_radix == 16 && base < 34)
-            {
-                return None;
             }
+            16 => lit.get(1).map_or(false, |&b| matches!(b, b'x' | b'X')),
+            2 => lit.get(1).map_or(false, |&b| matches!(b, b'b' | b'B')),
+            8 => lit.get(1).map_or(false, |&b| matches!(b, b'o' | b'O')),
+            _ => false,
+        }
+    } else {
+        if base == 0 {
+            base = 10;
+        }
+        false
+    };
+    if has_radix {
+        lit = &lit[2..];
+        if lit.first() == Some(&b'_') {
+            lit = &lit[1..];
         }
     }
 
-    // base still not found, try to use default
-    if base == 0 {
-        if buf.starts_with('0') {
-            if buf.chars().all(|c| matches!(c, '+' | '-' | '0' | '_')) {
-                return Some(BigInt::zero());
+    // remove zeroes
+    let mut last = *lit.first()?;
+    if last == b'0' {
+        let mut count = 0;
+        for pair in lit.windows(2) {
+            // get_unchecked below is safe because windows guarantees
+            let c1 = unsafe { pair.get_unchecked(0) };
+            let c2 = unsafe { pair.get_unchecked(1) };
```

Comment:
not sure if this is necessary, it should be able to optimize away bounds checks for stuff like `windows()` and `chunks_exact()`:
```rust
fn foo(x: &[u8]) {
    for win in x.windows(2) {
        print(win[0]);
        print(win[1]);
    }
}
```
```asm
foo:
        push    r14
        push    rbx
        push    rax
        cmp     rsi, 2
        jb      .LBB0_3
        mov     r14, rsi
        mov     rbx, rdi
        add     rbx, 1
.LBB0_2:
        add     r14, -1
        movzx   edi, byte ptr [rbx - 1]
        call    example::print
        movzx   edi, byte ptr [rbx]
        call    example::print
        add     rbx, 1
        cmp     r14, 2
        jae     .LBB0_2
.LBB0_3:
        add     rsp, 8
        pop     rbx
        pop     r14
        ret
```
https://godbolt.org/z/YvxPrh

Reply:
[playground link](https://play.rust-lang.org/?version=stable&mode=release&edition=2018), hit "asm" under the "..." next to "Build"; there's a bounds check for
```rust
        if lit.first() == Some(&b'_') {
            lit = &lit[1..]; // <<<<
        }
```
```rust
        // get_unchecked below is safe because count < lit.len() here
        let prefix_last = unsafe { *lit.get_unchecked(count) };
        lit = &lit[count + 1..]; // <<<<
```

Reply:
the link doesn't include code (share link?), but I got what you meant. that's unfortune.

Reply:
Oh, lol, I'm an idiot. https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=d1eb4f68631167b6155c0b149f96825b

Reply:
I figured out the line numbers by looking at the data sections in the bottom, and then the ones that referenced `L__unnamed_8` ("src/lib.rs") had a non-zero byte in the middle (`>` in `"\n\000\000\000\000\000\000\000>\000\000\000\024\000\000"`), which was the line number (`ord('>') == 62 and ord('Z') == 90`). Note that when I removed the get_unchecked on line 89, there was still just one bounds check because of the line after.

Reply:
I have no idea how did you read bound check from there. Does bound check included in "src/lib.rs"? In that case, I wish this is better:
https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=b89d48f636eb6989dcfd5b8b8ee8910c

Reply:
There are a few labels (e.g. `.LBB0_65`) near the bottom that places in the rest of the code jump to when a bounds check fails; those reference data labels at the very bottom like `.L__unnamed_5`:
```asm
.L__unnamed_8:
	.ascii	"src/lib.rs"

.L__unnamed_1:
	.byte	48

.L__unnamed_2:
	.byte	95

.L__unnamed_6:
	.quad	.L__unnamed_8
	.asciz	"\n\000\000\000\000\000\000\000>\000\000\000\024\000\000"
;                                              ^ lineno     ^^^^ column, i think

.L__unnamed_5:
	.quad	.L__unnamed_8
	.asciz	"\n\000\000\000\000\000\000\000`\000\000\000\033\000\000"
;                                              ^ lineno     ^^^^ column, i think
```
I know that that panic location is in our code because they reference `.L__unnamed_8` (src/lib.rs), and I figured out the location from the byte in the middle of the data string.